### PR TITLE
[refactor] AsyncPipe: do not sub-class MultiProcessPipe

### PR DESCRIPTION
**Pull request by msbaines**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by msbaines** [comment]
ping

File: fairscale/nn/pipe/async_pipeline.py [hunk 1]
```diff
@@ -4,18 +4,54 @@
...
+from .async_schedule import AsyncEventLoop, ModuleWrapper
+from .messages import MakeTransport
 from .microbatch import Batch
-from .multiprocess_pipeline import MultiProcessPipeline
+from .skip.layout import SkipLayout
 from .skip.tracker import SkipTrackerThroughPotals
 
 
-class AsyncPipeline(MultiProcessPipeline):
+class AsyncPipeline:
```
Comment (anj-s) [inline fairscale/nn/pipe/async_pipeline.py:20]: Why do we not need to subclass this anymore?

File: fairscale/nn/pipe/async_pipe.py [hunk 2]
```diff
@@ -38,10 +44,171 @@ def __len__(self) -> int:
...
+    """Wraps an arbitrary :class:`nn.Sequential <torch.nn.Sequential>` module
+    to train on Pipe_. If the module requires lots of memory, Pipe will be
+    very efficient.
+    ::
+
+        model = nn.Sequential(a, b, c, d)
+        model = Pipe(model, balance=[1, 1, 1, 1], chunks=8)
+        output = model(input)
+
+    .. _Pipe: https://arxiv.org/abs/1811.06965
```
Comment (anj-s) [inline fairscale/nn/pipe/async_pipe.py:75]: Would be good to have this docstring mention what is different between MultiprocessPipe and AsyncPipe? I don't know if this doc string is relevant anymore.

File: fairscale/nn/pipe/async_pipe.py [hunk 3]
```diff
@@ -150,3 +317,85 @@ def append_module(mod: "OrderedDict[str, nn.Module]") -> None:
...
+
+        raise IndexError
+
+    def __iter__(self) -> Iterable[nn.Module]:
+        """Iterates over children of the underlying sequential module."""
+        for partition in self.partitions:
+            yield from partition.module
+
+    def forward(self, input: TensorOrTensors, *, event=None) -> TensorOrTensors:  # type: ignore
+        """:class:`MultiProcessPipe` is a fairly transparent module wrapper. It doesn't
```
Comment (anj-s) [inline fairscale/nn/pipe/async_pipe.py:357]: change to Async?

File: fairscale/nn/pipe/async_pipeline.py (hunk 1 above)
Comment (msbaines) [reply fairscale/nn/pipe/async_pipeline.py:20]: I want to break the cyclic dependency between multiprocess_pipeline and async_schedule. The class hierarchy is a bit of a mess currently. I also don't like that the run function of *Pipeline takes a different argument type in the two classes.

Removing the inheritance should enables implication of MultiProcess pipe in order to understand what is really going on. I may add the inheritance back later, or add a common interface base-class, if it makes sense but right now it just gets in the way.

**Review by anj-s — APPROVED** [review]

**Review by sidgoyal78 — APPROVED** [review]
Looks good. Thanks for the PR.

**Pull request merged**